In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

### 3A.1. О чём этот модуль

# План модуля:
* Линейная регрессия.

### 3A.2. Линейная регрессия

In [10]:
myData = pd.read_csv('mycar.csv')

In [12]:
myData

Speed  Stopping_dist
0       4              2
1       4             10
2       7              4
3       7             22
4       8             16
5       9             10
6      10             18
7      10             26
8      10             34
9      11             17
10     11             28
11     12             14
12     12             20
13     12             24
14     12             28
15     13             26
16     13             34
17     13             34
18     13             46
19     14             26
20     14             36
21     14             60
22     14             80
23     15             20
24     15             26
25     15             54
26     16             32
27     16             40
28     17             32
29     17             40
30     17             50
31     18             42
32     18             56
33     18             76
34     18             84
35     19             36
36     19             46
37     19             68
38     20             32
39     20             48
40     20             52
41     20             56
42     20             64
43     22             66
44     23             54
45     24             70
46     24             92
47     24             93
48     25             85

In [13]:
X = myData.iloc[:,:-1].values
Y = myData.iloc[:,1].values

In [14]:
from sklearn.model_selection import train_test_split

In [15]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size= 0.3)

In [16]:
from sklearn.linear_model import LinearRegression
myModel = LinearRegression() #Обозначаем, что наша модель - линейная регрессия
myModel.fit(X_train,Y_train) #обучаем модель на обучающих данных

LinearRegression()

In [21]:
y_pred = myModel.predict(X_test)
y_pred

array([28.64662837, 16.1733689 , 13.05505403, 47.35651758, 50.47483244,
       47.35651758, 47.35651758, 31.76494324, 69.18472165, 66.06640678,
       53.59314731, 34.8832581 , 28.64662837, 22.40999863, 22.40999863])

### Задание 3.2.1
### Какая из этих задач является задачей линейной регрессии?
Предсказание зарплаты на основе опыта работы, образования и т. д. 

### Задание 3.2.2
### Почему для этих данных не подходит модель линейной регрессии?
Дисперсия шума зависит от x

### Задание 3.2.3
### Какой размер будет у матрицы X (матрица признаков) для данных в таблице? Задача — модель предсказания продаж.
(3*2)

### 3A.3. Ошибки в линейной регрессии

In [61]:
import pandas as pd
import numpy as np

In [62]:
y_happy = [1, 3, 2, 5]

In [63]:
y_happy

[1, 3, 2, 5]

In [64]:
y_happy_pred = [2, 3, -1, 4]

In [65]:
y_happy_pred

[2, 3, -1, 4]

In [66]:
from sklearn import metrics

In [67]:
MAE = metrics.mean_absolute_error(y_happy,y_happy_pred)

print(MAE)

1.25


In [68]:
MSE = metrics.mean_squared_error(y_happy,y_happy_pred)
print(MSE)

2.75


In [69]:
R_2 = metrics.r2_score(y_happy,y_happy_pred)
print(R_2)

-0.2571428571428571


### Задание 3.3.1
### Чему равна MSE на этих данных *F*?
 * 2.75

### Задание 3.3.2
### Что из этого является причиной использования MSE для задачи линейной регрессии?
* MSE сильнее штрафует за большие ошибки


### Задание 3.3.3
### У какой из этих моделей большой bias, а у какой — variance?
* У левой большой variance
* У правой большой bias

### 3A.4. Поиск прямой

### Задание 3.4.1
### Для каких данных лучше использовать *OLS*?

### Задание 3.4.2

### Задание 3.4.3

### Задание 3.4.4

### Задание 3.4.5

### Задание 3.4.6